### PR TITLE
Implement caching for GenerateContainerIDFromOriginInfo

### PR DESCRIPTION
**Pull request by wdhif**
### What does this PR do?

Implement caching for `GenerateContainerIDFromOriginInfo` using the `github.com/DataDog/datadog-agent/pkg/util/cache` implementation.

### Motivation

This is needed to fully support `GenerateContainerIDFromOriginInfo` use cases.

### Describe how you validated your changes

Validated using the following patch:
```
diff --git a/comp/core/tagger/impl-remote/remote.go b/comp/core/tagger/impl-remote/remote.go
index 7dd668b18a..da9c5ed25e 100644
--- a/comp/core/tagger/impl-remote/remote.go
+++ b/comp/core/tagger/impl-remote/remote.go
@@ -45,7 +45,7 @@ import (
 const (
        noTimeout         = 0 * time.Minute
        streamRecvTimeout = 10 * time.Minute
-       cacheExpiration   = 1 * time.Minute
+       cacheExpiration   = 10 * time.Second
 )

 var (
diff --git a/pkg/util/cache/cache.go b/pkg/util/cache/cache.go
index e78a1562fd..f2c751d12a 100644
--- a/pkg/util/cache/cache.go
+++ b/pkg/util/cache/cache.go
@@ -7,6 +7,7 @@
 package cache

 import (
+       "github.com/DataDog/datadog-agent/pkg/util/log"
        "strings"
        "time"

@@ -88,8 +89,10 @@ func Get[T any](key string, cb func() (T, error)) (T, error) {
 //     cached with the given expire duration and returned.
 func GetWithExpiration[T any](key string, cb func() (T, error), expire time.Duration) (T, error) {
        if x, found := Cache.Get(key); found {
+               log.Errorf("w: hit for key %s", key)
                return x.(T), nil
        }
+       log.Errorf("w: miss for key %s", key)

        res, err := cb()
        // We don't cache errors
```

And with the same QA process as https://github.com/DataDog/datadog-agent/pull/32295
We can see the logs:
```
2024-12-19 09:59:01 UTC | TRACE | ERROR | (pkg/util/cache/cache.go:92 in ]) | w: hit for key agent/remoteTagger/container_id/7c61e76f-5e3c-4989-b479-b977dd9cfe48/dd-trace-py
2024-12-19 09:59:02 UTC | TRACE | ERROR | (pkg/util/cache/cache.go:92 in ]) | w: hit for key agent/remoteTagger/container_id/7c61e76f-5e3c-4989-b479-b977dd9cfe48/dd-trace-py
2024-12-19 09:59:02 UTC | TRACE | ERROR | (pkg/util/cache/cache.go:92 in ]) | w: hit for key agent/remoteTagger/container_id/7c61e76f-5e3c-4989-b479-b977dd9cfe48/dd-trace-py
2024-12-19 09:59:06 UTC | TRACE | ERROR | (pkg/util/cache/cache.go:92 in ]) | w: hit for key agent/remoteTagger/container_id/7c61e76f-5e3c-4989-b479-b977dd9cfe48/dd-trace-py
2024-12-19 09:59:07 UTC | TRACE | ERROR | (pkg/util/cache/cache.go:95 in ]) | w: miss for key agent/remoteTagger/container_id/7c61e76f-5e3c-4989-b479-b977dd9cfe48/dd-trace-py
2024-12-19 09:59:10 UTC | TRACE | ERROR | (pkg/util/cache/cache.go:92 in ]) | w: hit for key agent/remoteTagger/container_id/7c61e76f-5e3c-4989-b479-b977dd9cfe48/dd-trace-py
2024-12-19 09:59:12 UTC | TRACE | ERROR | (pkg/util/cache/cache.go:92 in ]) | w: hit for key agent/remoteTagger/container_id/7c61e76f-5e3c-4989-b479-b977dd9cfe48/dd-trace-py
2024-12-19 09:59:14 UTC | TRACE | ERROR | (pkg/util/cache/cache.go:92 in ]) | w: hit for key agent/remoteTagger/container_id/7c61e76f-5e3c-4989-b479-b977dd9cfe48/dd-trace-py
```

And that we still have the traces correctly tagged:
image

### Possible Drawbacks / Trade-offs

N/A

### Additional Notes

N/A

### CODE REVIEW

**Review by jeremy-hanna — APPROVED** [review]
👍

File: comp/core/tagger/impl-remote/remote.go [hunk 1]
```diff
@@ -269,65 +272,78 @@ func (t *remoteTagger) GenerateContainerIDFromOriginInfo(originInfo origindetect
...
+		initPrefix+originInfo.ExternalData.PodUID+"/"+originInfo.ExternalData.ContainerName,
+	)
 
-	err := backoff.Retry(func() error {
-		select {
-		case <-t.ctx.Done():
-			return &backoff.PermanentError{Err: errTaggerFailedGenerateContainerIDFromOriginInfo}
-		default:
-		}
+	cachedContainerID, err := cache.GetWithExpiration(key, func() (containerID string, err error) {
```
Comment (clamoriniere) [inline comp/core/tagger/impl-remote/remote.go:286]: for code clarity we define the function outside the call of `cache.GetWithExpiration()` 
I feel that we have to many nesting here.

File: comp/core/tagger/impl-remote/remote.go [hunk 2]
```diff
@@ -269,65 +272,78 @@ func (t *remoteTagger) GenerateContainerIDFromOriginInfo(originInfo origindetect
...
+		expBackoff := backoff.NewExponentialBackOff()
+		expBackoff.InitialInterval = 500 * time.Millisecond
```
Comment (clamoriniere) [inline comp/core/tagger/impl-remote/remote.go:288]: maybe we can start by 200 instead of 500

Comment (wdhif) [reply comp/core/tagger/impl-remote/remote.go]: Done 👍

File: comp/core/tagger/impl-remote/remote.go (hunk 1 above)
Comment (wdhif) [reply comp/core/tagger/impl-remote/remote.go:286]: Sadly this does not seem possible without modifying the `pkg/util/cache/cache`, which I would prefer avoid doing 👍

**Review by gabedos — APPROVED** [review]
LGTM. Unfortunate we couldn't extract the callback func out.

File: comp/core/tagger/impl-remote/remote.go [hunk 3]
```diff
@@ -269,65 +272,78 @@ func (t *remoteTagger) GenerateContainerIDFromOriginInfo(originInfo origindetect
...
-		}
 
-		// Create the context with the auth token
-		t.queryCtx, t.queryCancel = context.WithCancel(
-			metadata.NewOutgoingContext(t.ctx, metadata.MD{
-				"authorization": []string{fmt.Sprintf("Bearer %s", t.token)},
-			}),
-		)
+			// Create the context with the auth token
+			t.queryCtx, t.queryCancel = context.WithCancel(
```
Comment (clamoriniere) [inline comp/core/tagger/impl-remote/remote.go:310]: I think we have a race here if several goroutine cal the function `GenerateContainerIDFromOriginInfo` 

`queryCtx` and `queryCancel` need to be local and not part of the `remoteTagger` struct

File: comp/core/tagger/impl-remote/remote.go (hunk 1 above)
Comment (wdhif) [reply comp/core/tagger/impl-remote/remote.go:286]: We checked together, all good 👍

File: comp/core/tagger/impl-remote/remote.go (hunk 3 above)
Comment (wdhif) [reply comp/core/tagger/impl-remote/remote.go:310]: Agreed, done 👍

**Review by clamoriniere — APPROVED** [review]

**Pull request merged**